File: 1 - LinkedList/C#/4-CircularDoubleLinkedList/circularSingleLinkedList.cs
using System;

namespace CircularDoubleLinkedList
{
    // Node class representing each element in the circular doubly linked list
    public class Node
    {
        public string Name { get; set; }     // Item name
        public int Price { get; set; }       // Item price
        public int Available { get; set; }   // Item availability
        public Node Prev { get; set; }       // Pointer to the previous node in the list
        public Node Next { get; set; }       // Pointer to the next node in the list

        // Constructor to initialize a new node
        public Node(string name, int price, int available)
        {
            Name = name;
            Price = price;
            Available = available;
            Prev = null;
            Next = null;
        }
    }

    // LinkedList class to manage the circular doubly linked list
    public class LinkedList
    {
        private Node head; // Pointer to the head of the list
        private Node tail; // Pointer to the tail of the list
        private int size;  // Size of the list

        // Constructor to initialize an empty list
        public LinkedList()
        {
            head = null;
            tail = null;
            size = 0;
        }

        // Function to create a circular doubly linked list with the first node
        public void CreateCircularDL(string name, int price, int available)
        {
            Node newNode = new Node(name, price, available);
            newNode.Next = newNode;
            newNode.Prev = newNode;
            head = newNode;
            tail = newNode;
            size = 1;
        }

        // Function to add a node at the beginning of the list
        public void AddAtFirstNode(string name, int price, int available)
        {
            if (size == 0)
            {
                CreateCircularDL(name, price, available);
                return;
            }
            Node newNode = new Node(name, price, available);
            newNode.Next = head;
            newNode.Prev = tail;
            head.Prev = newNode;
            tail.Next = newNode;
            head = newNode;
            size++;
        }

        // Function to add a node at the nth position in the list
        public void AddAtNthNode(string name, int price, int available, int pos)
        {
            if (pos < 1 || pos > size + 1)
            {
                Console.WriteLine("Invalid position!");
                return;
            }
            if (pos == 1)
            {
                AddAtFirstNode(name, price, available);
                return;
            }
            if (pos == size + 1)
            {
                AddAtLastNode(name, price, available);
                return;
            }
            Node newNode = new Node(name, price, available);
            Node cur = head;
            for (int i = 1; i < pos - 1; i++)
            {
                cur = cur.Next;
            }
            newNode.Next = cur.Next;
            newNode.Prev = cur;
            cur.Next.Prev = newNode;
            cur.Next = newNode;
            size++;
        }

        // Function to add a node at the end of the list
        public void AddAtLastNode(string name, int price, int available)
        {
            if (size == 0)
            {
                CreateCircularDL(name, price, available);
                return;
            }
            Node newNode = new Node(name, price, available);
            newNode.Next = head;
            newNode.Prev = tail;
            tail.Next = newNode;
            head.Prev = newNode;
            tail = newNode;
            size++;
        }

        // Function to update the first node in the list
        public void UpdateFirstNode(string name, int price, int available)
        {
            if (size == 0)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            head.Name = name;
            head.Price = price;
            head.Available = available;
        }

        // Function to update the nth node in the list
        public void UpdateNthNode(string name, int price, int available, int pos)
        {
            if (pos < 1 || pos > size)
            {
                Console.WriteLine("Invalid position!");
                return;
            }
            Node cur = head;
            for (int i = 1; i < pos; i++)
            {
                cur = cur.Next;
            }
            cur.Name = name;
            cur.Price = price;
            cur.Available = available;
        }

        // Function to update the last node in the list
        public void UpdateLastNode(string name, int price, int available)
        {
            if (size == 0)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            tail.Name = name;
            tail.Price = price;
            tail.Available = available;
        }

        // Function to remove the first node from the list
        public void RemoveFirstNode()
        {
            if (size == 0)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            Node temp = head;
            if (size == 1)
            {
                head = null;
                tail = null;
            }
            else
            {
                head = head.Next;
                head.Prev = tail;
                tail.Next = head;
            }
            size--;
        }

        // Function to remove the nth node from the list
        public void RemoveNthNode(int pos)
        {
            if (pos < 1 || pos > size)
            {
                Console.WriteLine("Invalid position!");
                return;
            }
            if (pos == 1)
            {
                RemoveFirstNode();
                return;
            }
            if (pos == size)
            {
                RemoveLastNode();
                return;
            }
            Node cur = head;
            for (int i = 1; i < pos; i++)
            {
                cur = cur.Next;
            }
            cur.Prev.Next = cur.Next;
            cur.Next.Prev = cur.Prev;
            size--;
        }

        // Function to remove the last node from the list
        public void RemoveLastNode()
        {
            if (size == 0)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            Node temp = tail;
            if (size == 1)
            {
                head = null;
                tail = null;
            }
            else
            {
                tail = tail.Prev;
                tail.Next = head;
                head.Prev = tail;
            }
            size--;
        }

        // Function to print the circular doubly linked list
        public void PrintCircularDL()
        {
            if (size == 0)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            Node cur = head;
            do
            {
                Console.WriteLine($"Name: {cur.Name}, Price: {cur.Price}, Available: {cur.Available}");
                cur = cur.Next;
            } while (cur != head);
        }

        // Utility function to count the number of nodes in the list
        public int CountNodes()
        {
            return size; // Return the total number of nodes
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            LinkedList list = new LinkedList();

            // Create the circular doubly linked list with the first node
            list.CreateCircularDL("Laptop", 20000000, 20);
            list.PrintCircularDL();
            Console.WriteLine("==========\n\n\n");

            // Add a node at the beginning of the list
            list.AddAtFirstNode("Handphone", 5000000, 50);
            list.PrintCircularDL();
            Console.WriteLine("==========\n\n\n");

            // Add nodes at the end of the list
            list.AddAtLastNode("Mouse", 1200000, 100);
            list.AddAtLastNode("Keyboard", 2250000, 100);
            list.AddAtLastNode("Headphone", 1800000, 100);
            list.AddAtLastNode("Monitor", 4200000, 100);
            list.PrintCircularDL();
            Console.WriteLine("==========\n\n\n");

            // Remove the first node from the list
            list.RemoveFirstNode();
            list.PrintCircularDL();
            Console.WriteLine("==========\n\n\n");

            // Remove the last node from the list
            list.RemoveLastNode();
            list.PrintCircularDL();
            Console.WriteLine("==========\n\n\n");

            // Add a node at the nth position in the list
            list.AddAtNthNode("RTX 5090", 40000000, 15, 3);
            list.PrintCircularDL();
            Console.WriteLine("==========\n\n\n");

            // Remove the nth node from the list
            list.RemoveNthNode(4);
            list.PrintCircularDL();
            Console.WriteLine("==========\n\n\n");

            // Update the first node in the list
            list.UpdateFirstNode("Laptop", 19000000, 10);
            // Update the nth node in the list
            list.UpdateNthNode("Mouse", 1000000, 3, 2);
            // Update the last node in the list
            list.UpdateLastNode("Headphone", 1900000, 5);
            list.PrintCircularDL();
            Console.WriteLine("==========\n\n\n");

            // Print the total number of nodes in the list
            Console.WriteLine("Total nodes: " + list.CountNodes());
        }
    }
}

File: 1 - LinkedList/Php/3-CircularSingleLinkedList/circularSingleLinkedList.php
<?php

/**
 * Class representing a node in the circular singly linked list.
 */
class Node {
    public $id;
    public $name;
    public $major;
    public $next;

    /**
     * Create a node.
     * @param string $id - The ID of the data.
     * @param string $name - The name of the data.
     * @param string $major - The major associated with the data.
     */
    public function __construct($id, $name, $major) {
        $this->id = $id;
        $this->name = $name;
        $this->major = $major;
        $this->next = null; // Pointer to the next node
    }
}

/**
 * Class representing a circular singly linked list.
 */
class LinkedList {
    private $head;
    private $tail;
    private $size;

    /**
     * Create a linked list.
     */
    public function __construct() {
        $this->head = null; // Pointer to the head of the list
        $this->tail = null; // Pointer to the tail of the list
        $this->size = 0; // Size of the list
    }

    /**
     * Create a circular singly linked list with the first node.
     * @param string $id - The ID of the data.
     * @param string $name - The name of the data.
     * @param string $major - The major associated with the data.
     */
    public function createCircularSL($id, $name, $major) {
        $newNode = new Node($id, $name, $major);
        $newNode->next = $newNode;
        $this->head = $newNode;
        $this->tail = $newNode;
        $this->size = 1;
    }

    /**
     * Add a node at the beginning of the list.
     * @param string $id - The ID of the data.
     * @param string $name - The name of the data.
     * @param string $major - The major associated with the data.
     */
    public function addAtFirstNode($id, $name, $major) {
        if ($this->size === 0) {
            $this->createCircularSL($id, $name, $major);
            return;
        }
        $newNode = new Node($id, $name, $major);
        $newNode->next = $this->head;
        $this->tail->next = $newNode;
        $this->head = $newNode;
        $this->size++;
    }

    /**
     * Add a node at the nth position in the list.
     * @param string $id - The ID of the data.
     * @param string $name - The name of the data.
     * @param string $major - The major associated with the data.
     * @param int $pos - The position to insert the node.
     */
    public function addAtNthNode($id, $name, $major, $pos) {
        if ($pos < 1 || $pos > $this->size + 1) {
            echo "Invalid position!\n";
            return;
        }
        if ($pos == 1) {
            $this->addAtFirstNode($id, $name, $major);
            return;
        }
        if ($pos == $this->size + 1) {
            $this->addAtLastNode($id, $name, $major);
            return;
        }
        $newNode = new Node($id, $name, $major);
        $current = $this->head;
        for ($i = 1; $i < $pos - 1; $i++) {
            $current = $current->next;
        }
        $newNode->next = $current->next;
        $current->next = $newNode;
        $this->size++;
    }

    /**
     * Add a node at the end of the list.
     * @param string $id - The ID of the data.
     * @param string $name - The name of the data.
     * @param string $major - The major associated with the data.
     */
    public function addAtLastNode($id, $name, $major) {
        if ($this->size === 0) {
            $this->createCircularSL($id, $name, $major);
            return;
        }
        $newNode = new Node($id, $name, $major);
        $newNode->next = $this->head;
        $this->tail->next = $newNode;
        $this->tail = $newNode;
        $this->size++;
    }

    /**
     * Update the first node in the list.
     * @param string $id - The ID of the data.
     * @param string $name - The name of the data.
     * @param string $major - The major associated with the data.
     */
    public function updateFirstNode($id, $name, $major) {
        if ($this->size === 0) {
            echo "List is empty!\n";
            return;
        }
        $this->head->id = $id;
        $this->head->name = $name;
        $this->head->major = $major;
    }

    /**
     * Update the nth node in the list.
     * @param string $id - The ID of the data.
     * @param string $name - The name of the data.
     * @param string $major - The major associated with the data.
     * @param int $pos - The position of the node to update.
     */
    public function updateNthNode($id, $name, $major, $pos) {
        if ($pos < 1 || $pos > $this->size) {
            echo "Invalid position!\n";
            return;
        }
        $current = $this->head;
        for ($i = 1; $i < $pos; $i++) {
            $current = $current->next;
        }
        $current->id = $id;
        $current->name = $name;
        $current->major = $major;
    }

    /**
     * Update the last node in the list.
     * @param string $id - The ID of the data.
     * @param string $name - The name of the data.
     * @param string $major - The major associated with the data.
     */
    public function updateLastNode($id, $name, $major) {
        if ($this->size === 0) {
            echo "List is empty!\n";
            return;
        }
        $this->tail->id = $id;
        $this->tail->name = $name;
        $this->tail->major = $major;
    }

    /**
     * Remove the first node from the list.
     */
    public function removeFirstNode() {
        if ($this->size === 0) {
            echo "List is empty!\n";
            return;
        }
        $del = $this->head;
        if ($this->size === 1) {
            $this->head = null;
            $this->tail = null;
        } else {
            $this->head = $this->head->next;
            $this->tail->next = $this->head;
        }
        $this->size--;
    }

    /**
     * Remove the nth node from the list.
     * @param int $pos - The position of the node to remove.
     */
    public function removeNthNode($pos) {
        if ($pos < 1 || $pos > $this->size) {
            echo "Invalid position!\n";
            return;
        }
        if ($pos == 1) {
            $this->removeFirstNode();
            return;
        }
        $current = $this->head;
        for ($i = 1; $i < $pos - 1; $i++) {
            $current = $current->next;
        }
        $del = $current->next;
        $current->next = $del->next;
        if ($del == $this->tail) {
            $this->tail = $current;
        }
        $this->size--;
    }

    /**
     * Remove the last node from the list.
     */
    public function removeLastNode() {
        if ($this->size === 0) {
            echo "List is empty!\n";
            return;
        }
        $del = $this->tail;
        if ($this->size === 1) {
            $this->head = null;
            $this->tail = null;
        } else {
            $current = $this->head;
            while ($current->next !== $this->tail) {
                $current = $current->next;
            }
            $this->tail = $current;
            $this->tail->next = $this->head;
        }
        $this->size--;
    }

    /**
     * Print all nodes in the circular singly linked list.
     */
    public function printCircularSL() {
        if ($this->size === 0) {
            echo "List is empty!\n";
            return;
        }
        echo "Number of Nodes: " . $this->size . "\n";
        $current = $this->head;
        $i = 1;
        do {
            echo "Node " . $i . "\n";
            echo "ID\t\t: " . $current->id . "\n";
            echo "Name\t\t: " . $current->name . "\n";
            echo "Major\t\t: " . $current->major . "\n\n";
            $current = $current->next;
            $i++;
        } while ($current !== $this->head);
    }

    /**
     * Utility function to count the number of nodes in the list.
     * @return int The total number of nodes.
     */
    public function countNodes() {
        return $this->size;
    }
}

// Example usage
$list = new LinkedList();
$list->createCircularSL("11223399", "Arief F-sa", "Computer Science");
$list->printCircularSL();
echo "==========\n\n";

$list->addAtFirstNode("11223398", "Nisa Uswa", "Psychology");
$list->printCircularSL();
echo "==========\n\n";

$list->addAtLastNode("11223400", "Pejeel", "Drop out");
$list->printCircularSL();
echo "==========\n\n";

$list->removeFirstNode();
$list->printCircularSL();
echo "==========\n\n";

$list->removeLastNode();
$list->printCircularSL();
echo "==========\n\n";

$list->addAtFirstNode("11223398", "Nisa Uswa", "Psychology");
$list->addAtLastNode("11223400", "Pejeel", "Drop out");
$list->printCircularSL();
echo "==========\n\n";

$list->addAtNthNode("11223401", "dummy1", "Math", 3);
$list->addAtNthNode("11223402", "dummy2", "Physics", 2);
$list->printCircularSL();
echo "==========\n\n";

$list->removeNthNode(2);
$list->printCircularSL();
echo "==========\n\n";
?>